osu.Game/Beatmaps/Drawables/Cards/BeatmapCardDownloadProgressBar.cs

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Shapes;
+using osu.Game.Graphics;
+using osu.Game.Online;
+using osu.Game.Overlays;
+
+namespace osu.Game.Beatmaps.Drawables.Cards
+{
+    public class BeatmapCardDownloadProgressBar : CompositeDrawable
+    {
+        public override bool IsPresent => true;
+
+        private readonly CircularContainer foreground;
+
+        private readonly Box backgroundFill;
+        private readonly Box foregroundFill;
+
+        private readonly Bindable<DownloadState> state = new Bindable<DownloadState>();
+        private readonly BindableDouble progress = new BindableDouble();
+
+        [Resolved]
+        private OsuColour colours { get; set; }
+
+        [Resolved]
+        private OverlayColourProvider colourProvider { get; set; }
+
+        public BeatmapCardDownloadProgressBar()
+        {
+            InternalChildren = new Drawable[]
+            {
+                new CircularContainer
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Masking = true,
+                    Child = backgroundFill = new Box
+                    {
+                        RelativeSizeAxes = Axes.Both,
+                    }
+                },
+                foreground = new CircularContainer
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Masking = true,
+                    Child = foregroundFill = new Box
+                    {
+                        RelativeSizeAxes = Axes.Both,
+                    }
+                }
+            };
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(BeatmapDownloadTracker downloadTracker)
+        {
+            backgroundFill.Colour = colourProvider.Background6;
+
+            ((IBindable<DownloadState>)state).BindTo(downloadTracker.State);
```

Comment:
Not a huge fan of this being cached/resolved. Rationale is that in generally I don't like using DI for one-level-deep components - it feels too far-reaching for such usages (harder to follow the chain of data), especially for cases like buttons and UI controls.

I'd suggest either having each component create their own tracker, or binding inwards from the parent for the required bindables (state/progress as required).

Reply:
Separate trackers is how I've had it initially, but it didn't feel too great to me, so I went for binding from the parent card. Seems alright for me.